content/en/blog/_posts/2021-09-07-data-duplication-in-data-heavy-k8s-env.md

```diff
@@ -0,0 +1,221 @@
+---
+layout: blog
+title: "How to Handle Data Duplication in Data-Heavy Kubernetes Environments"
+date: 2021-09-07
+slug: how-to-handle-data-duplication-in-data-heavy-kubernetes-environments 
+---
+
+**Authors:**
+Augustinas Stirbis (CAST AI)
+
+# How to Handle Data Duplication in Data-Heavy Kubernetes Environments
+
+It’s convenient to create a copy of your application with a copy of its state for each team. 
+For example, you might want a separate database copy to test some significant schema changes 
+or develop other disruptive operations like bulk insert/delete/update...
+
+**Duplicating data takes a lot of time.** That’s because you need first to download 
+all the data from a source like object storage (S3 or Cloud Storage) and then send 
+it back to a storage back-end like EBS or Standard Persistent Disk. There’s a lot of network traffic and CPU/RAM used in this process.
+
+Hardware acceleration by offloading certain expensive operations is **always a huge 
+performance boost**. It reduces the time required to complete an operation by orders 
+of magnitude - for example, GPU computation while you’re watching a high-resolution 
+video on Youtube, mining crypto, or training an ML model, CPU instructions for AES 
+disk encryption or TLS, TCP offload (TOE), and many many more.
+
+## Volume Snapshots to the rescue
+
+Kubernetes introduced Volume Snapshots alpha in 1.12, beta in 1.17, and the Generally Available 
+version in 1.20. Volume Snapshots basically allows Kubernetes to use the cloud-native storage 
+snapshotting mechanism that allows data copy to offload from compute to the actual storage device.
+
+Since data is already in the same storage device (array of devices), duplicating data is usually 
+a metadata operation. All you need to do is point a new disk to an immutable snapshot and only 
+save deltas (or let it do a full-disk copy). As an operation that is inside the storage back-end,
+it’s much quicker and usually doesn’t involve sending traffic over the network.
+
+## Sounds like we have our solution, right?
+
+Actually, Volume Snapshots are namespaced, and Kubernetes protects namespaced data from 
+being shared between tenants (namespaces). This Kubernetes limitation is a conscious design decision so that a pod in a different namespace can’t mount another application’s Persistent Volume (Claim).
+
+One way around it would be to create duplicate data accessible on the same namespace. 
```

Comment:
Can you clarify what you mean by "create duplicate data accessible..."?

> You could create multiple volumes with duplicate data in one namespace. However, you could easily reference the wrong copy.

Reply:
Thank you for the suggestion, it's more clear now!